In [28]:
from pandas import *
import numpy as np
from scipy.stats.stats import pearsonr
import itertools

from datacontainer import *
import sys
np.set_printoptions(threshold=sys.maxsize)
#%Read the file in%


## Load Data

Load data and header from a dataframe.

In [29]:
datamatrix = DataContainer('102318 UNIQUE DRUGS 32 Mol Props.csv').dataMatrix
header = DataContainer('102318 UNIQUE DRUGS 32 Mol Props.csv').header
# make  a dataframe
df=DataFrame(np.array(datamatrix))


## Correlation Analysis of Features
Find the pairwise correlations for the feature


In [30]:

correlations = {}
columns = df.columns.tolist()

# find all combinations of columns
for col_a, col_b in itertools.combinations(columns, 2):
    correlations[str(col_a) + '__' + str(col_b) +  "/(%s" % header[col_a] +"-%s)" % header[col_b]] = pearsonr(df.loc[:, col_a], df.loc[:, col_b])

result = DataFrame.from_dict(correlations, orient='index')
result.columns = ['PCC', 'p-value']

# sort and display the correlations by PCC
print(result.sort_values('PCC',ascending=False))


                                             PCC       p-value
2__6/(molVolume-molArea)                0.982333  1.015966e-41
1__17/(molWeight-a_heavy)               0.974235  2.935002e-37
6__17/(molArea-a_heavy)                 0.971431  4.849183e-36
2__17/(molVolume-a_heavy)               0.968683  5.847983e-35
8__25/(nof_Atoms-nof_Fragments)         0.952817  3.735533e-30
1__2/(molWeight-molVolume)              0.945288  1.984079e-28
1__6/(molWeight-molArea)                0.940566  1.817579e-27
1__14/(molWeight-Complexity)            0.939141  3.421802e-27
6__25/(molArea-nof_Fragments)           0.936470  1.076309e-26
5__10/(molPSA-nof_HBA)                  0.915475  2.096237e-23
12__16/(molCharge_total-nof_negCharge)  0.913432  3.933105e-23
2__25/(molVolume-nof_Fragments)         0.901967  1.031966e-21
14__17/(Complexity-a_heavy)             0.901574  1.146121e-21
6__8/(molArea-nof_Atoms)                0.889156  2.545206e-20
2__14/(molVolume-Complexity)            0.887300  3.919

C:\Users\Kaustubh\Anaconda3\envs\deeplearning\lib\site-packages\scipy\stats\stats.py:3010: RuntimeWarning: invalid value encountered in double_scalars
  r = r_num / r_den
C:\Users\Kaustubh\Anaconda3\envs\deeplearning\lib\site-packages\scipy\stats\stats.py:5256: RuntimeWarning: invalid value encountered in less
  x = np.where(x < 1.0, x, 1.0)  # if x > 1 then return 1.0


`
Large Positive Correlations (>0.85):
                                             PCC       p-value
2__6/(molVolume-molArea)                0.982333  1.015966e-41
1__17/(molWeight-a_heavy)               0.974235  2.935002e-37
6__17/(molArea-a_heavy)                 0.971431  4.849183e-36
2__17/(molVolume-a_heavy)               0.968683  5.847983e-35
8__25/(nof_Atoms-nof_Fragments)         0.952817  3.735533e-30
1__2/(molWeight-molVolume)              0.945288  1.984079e-28
1__6/(molWeight-molArea)                0.940566  1.817579e-27
1__14/(molWeight-Complexity)            0.939141  3.421802e-27
6__25/(molArea-nof_Fragments)           0.936470  1.076309e-26
5__10/(molPSA-nof_HBA)                  0.915475  2.096237e-23
12__16/(molCharge_total-nof_negCharge)  0.913432  3.933105e-23
2__25/(molVolume-nof_Fragments)         0.901967  1.031966e-21
14__17/(Complexity-a_heavy)             0.901574  1.146121e-21
6__8/(molArea-nof_Atoms)                0.889156  2.545206e-20
2__14/(molVolume-Complexity)            0.887300  3.919000e-20
20__24/(C_R0-nof_RotB)                  0.882077  1.269471e-19
17__25/(a_heavy-nof_Fragments)          0.879822  2.072861e-19
19__23/(C_R1-negCharge/Volume)          0.872196  1.015024e-18
18__22/(C_R2-posCharge/Volume)          0.864518  4.547482e-18
Keep:  17, 25, 6, 1, 10, 16 + rest
Discard: 1, 2, 14, 5, 12, 18, 20, 23
`

## Delete Redundant Columns


In [31]:
todel = [1, 2, 14, 5, 12, 18, 20, 23]
df.drop(df.columns[todel], axis=1,inplace=True)

In [32]:
print('features removed')
print([(header[x], x) for x in sorted(todel)])
print()
print('kept', [header[x]for x in sorted([17, 25, 6, 1, 10, 16])])





features removed
[('molWeight', 1), ('molVolume', 2), ('molPSA', 5), ('molCharge_total', 12), ('Complexity', 14), ('C_R2', 18), ('C_R0', 20), ('negCharge/Volume', 23)]

kept ['molWeight', 'molArea', 'nof_HBA', 'nof_negCharge', 'a_heavy', 'nof_Fragments']
